sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala

```diff
@@ -351,7 +352,11 @@ abstract class OutputWriterFactory extends Serializable {
    *
    * @since 1.4.0
    */
-  def newInstance(path: String, dataSchema: StructType, context: TaskAttemptContext): OutputWriter
+  def newInstance(
+      path: String,
+      bucketId: Option[Int],
+      dataSchema: StructType,
+      context: TaskAttemptContext): OutputWriter
```

Comment:
We probably should overload `newInstance` here and fallback to the old one since `OutputWriterFactory.newInstance()` is part of the public data sources API (it's marked as experimental, though).